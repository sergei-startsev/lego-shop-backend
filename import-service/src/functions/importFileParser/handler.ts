import { Handler, S3EventRecord } from 'aws-lambda';
import { S3Client, GetObjectCommand, CopyObjectCommand, DeleteObjectCommand } from '@aws-sdk/client-s3';
import csv from 'csv-parser';
import { Readable } from 'stream';
import pino from 'pino';

const importFileParser: Handler = async event => {
  const logger = pino();
  logger.info('importFileParser');
  logger.info(event);

  const record = event?.Records?.find((r: S3EventRecord) => r.eventName === 'ObjectCreated:Put');
  if (!record) {
    return event;
  }

  const bucket = record.s3.bucket.name;
  const key = record.s3.object.key;

  logger.info({ bucket });
  logger.info({ key });

  const client = new S3Client({ region: process.env.REGION });
  const getCommand = new GetObjectCommand({
    Bucket: bucket,
    Key: key
  });

  const response = await client.send(getCommand);
  if (!response?.Body) {
    return event;
  }

  await new Promise<void>((resolve, reject) => {
    (response.Body as Readable)
      .pipe(csv({ strict: true }))
      .on('data', data => {
        logger.info(data);
      })
      .on('error', reject)
      .on('end', () => {
        logger.info('importFileParser finished');
        resolve();
      });
  });

  const newKey = key.replace(/^uploaded/, 'parsed');
  const copyCommand = new CopyObjectCommand({
    CopySource: `${bucket}/${key}`,
    Bucket: bucket,
    Key: newKey
  });
  await client.send(copyCommand);
  logger.info(`${key} -> ${newKey}`);

  const deleteCommand = new DeleteObjectCommand({
    Bucket: bucket,
    Key: key
  });
  await client.send(deleteCommand);
  logger.info(`${key} was deleted`);

  return event;
};

export const main = importFileParser;
