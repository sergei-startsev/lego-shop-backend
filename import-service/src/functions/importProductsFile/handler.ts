import { Handler } from 'aws-lambda';
import { PutObjectCommand, S3Client } from '@aws-sdk/client-s3';
import { getSignedUrl } from '@aws-sdk/s3-request-presigner';
import createError from 'http-errors';
import * as z from 'zod';
import pino from 'pino';
import { serializeResponse } from '@libs/api-gateway';
import { middyfy } from '@libs/lambda';

const importProductsFile: Handler = async event => {
  const logger = pino();
  logger.info('importProductsFile');
  logger.info(event);

  const schema = z.object({
    name: z.string().min(3)
  });
  const parsed = schema.safeParse(event?.queryStringParameters);
  if (!parsed.success) {
    throw createError(400, parsed.error);
  }
  const { name } = parsed.data;

  const client = new S3Client({ region: process.env.REGION });
  const command = new PutObjectCommand({
    Bucket: process.env.BUCKET_NAME,
    Key: `uploaded/${name}`
  });
  const url = await getSignedUrl(client, command, { expiresIn: 3600 });

  return serializeResponse({ url });
};

export const main = middyfy(importProductsFile);
