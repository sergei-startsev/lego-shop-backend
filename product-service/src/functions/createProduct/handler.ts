import { DynamoDBClient } from '@aws-sdk/client-dynamodb';
import { DynamoDBDocumentClient, TransactWriteCommand } from '@aws-sdk/lib-dynamodb';
import { Handler } from 'aws-lambda';
import createError from 'http-errors';
import { v4 as uuidv4 } from 'uuid';
import * as z from 'zod';
import pino from 'pino';
import { serializeResponse } from '@libs/api-gateway';
import { middyfy } from '@libs/lambda';
import { Product, Stock } from '@customTypes/api-types';

const createProduct: Handler = async event => {
  const logger = pino();
  logger.info('createProduct');
  logger.info(event);

  const schema = z.object({
    title: z.string(),
    description: z.string(),
    price: z.number().positive(),
    count: z.number().nonnegative()
  });

  const parsed = schema.safeParse(event.body);
  if (!parsed.success) {
    throw createError(400, parsed.error);
  }

  const { title, description, price, count } = parsed.data;
  const id = uuidv4();
  const product: Omit<Product, 'count'> = { id, title, description, price };
  const stock: Stock = { product_id: id, count };

  const client = new DynamoDBClient({ region: process.env.REGION });
  const ddbDocClient = DynamoDBDocumentClient.from(client);
  const createProduct = new TransactWriteCommand({
    TransactItems: [
      {
        Put: {
          TableName: process.env.PRODUCTS_TABLE_NAME,
          Item: product
        }
      },
      {
        Put: {
          TableName: process.env.STOCK_TABLE_NAME,
          Item: stock
        }
      }
    ]
  });
  await ddbDocClient.send(createProduct);

  const output: Product = { id, title, description, price, count };
  return serializeResponse(output);
};

export const main = middyfy(createProduct);
