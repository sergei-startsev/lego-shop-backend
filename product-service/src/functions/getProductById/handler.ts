import type { ValidatedEventAPIGatewayProxyEvent } from '@libs/api-gateway';
import { APIGatewayProxyEventV2 } from 'aws-lambda';
import { formatJSONResponse } from '@libs/api-gateway';
import mockData from '@mocks/data.json';

export type ValidatedAPIGatewayProxyEvent = Omit<APIGatewayProxyEventV2, 'pathParameters'> & {
  pathParameters: {
    id: string;
  };
};

const getProductById: ValidatedEventAPIGatewayProxyEvent<ValidatedAPIGatewayProxyEvent> = async event => {
  const id = event?.pathParameters?.id;
  if (!id) {
    return formatJSONResponse('Product ID not specified', 500);
  }
  const product = mockData.find(obj => obj.id === id);
  if (product) {
    return formatJSONResponse(product);
  } else {
    return formatJSONResponse('Product not found', 404);
  }
};

export const main = getProductById;
