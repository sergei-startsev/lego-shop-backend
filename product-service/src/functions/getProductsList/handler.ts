import type { ValidatedEventAPIGatewayProxyEvent } from '@libs/api-gateway';
import { formatJSONResponse } from '@libs/api-gateway';
import mockData from '@mocks/data.json';

const getProductsList: ValidatedEventAPIGatewayProxyEvent<void> = async () => formatJSONResponse(mockData);

export const main = getProductsList;
