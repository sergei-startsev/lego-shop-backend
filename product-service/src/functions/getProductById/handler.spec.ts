import { Context } from 'aws-lambda';
import { main } from './handler';

const buildEvent = (id: string) => ({
  version: '2.0',
  routeKey: 'GET /products/{id}',
  rawPath: `/products/${id}`,
  rawQueryString: '',
  headers: {
    accept: 'text/html',
    'accept-encoding': 'gzip, deflate, br'
  },
  requestContext: {
    accountId: '1',
    apiId: '1',
    domainName: 'foo.execute-api.eu-central-1.amazonaws.com',
    domainPrefix: 'foo',
    http: {
      method: 'GET',
      path: `/products/${id}`,
      protocol: 'HTTP/1.1',
      sourceIp: '127.0.0.1',
      userAgent: 'Chrome/110.0.0.0'
    },
    requestId: '111',
    routeKey: 'GET /products/{id}',
    stage: '$default',
    time: '20/Feb/2023:23:46:42 +0000',
    timeEpoch: 1676936802838
  },
  pathParameters: { id },
  isBase64Encoded: false
});

describe('getProductById', () => {
  it('should return product by id', async () => {
    const event = buildEvent('42133');
    const res = await main(event, {} as Context, () => {});
    expect(res).toMatchSnapshot();
  });

  it('should return 404 status code', async () => {
    const event = buildEvent('11111');
    const res = await main(event, {} as Context, () => {});
    expect(res).toMatchSnapshot();
  });

  it('should return 500 status code', async () => {
    const event = buildEvent('');
    const res = await main(event, {} as Context, () => {});
    expect(res).toMatchSnapshot();
  });
});
