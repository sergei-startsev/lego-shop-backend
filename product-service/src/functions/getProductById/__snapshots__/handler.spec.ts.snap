// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getProductById should return 404 status code 1`] = `
{
  "body": ""Product not found"",
  "statusCode": 404,
}
`;

exports[`getProductById should return 500 status code 1`] = `
{
  "body": ""Product ID not specified"",
  "statusCode": 500,
}
`;

exports[`getProductById should return product by id 1`] = `
{
  "body": "{"description":"This 2-in-1 model comes with a buildable pallet to lift, so kids can role-play different construction site adventures","id":"42133","price":50,"title":"Telehandler","count":25}",
  "statusCode": 200,
}
`;
